pypesto/optimize/ess.py

```diff
@@ -0,0 +1,662 @@
+            logger.debug("Local search only from best point.")
+            local_search_x0 = self.x_best
+            local_search_fx0 = self.fx_best
+        # first local search?
+        elif (
+            not self.local_solutions and self.evaluator.n_eval >= self.local_n1
+        ):
+            logger.debug(
+                "First local search from best point due to "
+                f"local_n1={self.local_n1}."
+            )
+            local_search_x0 = self.x_best
+            local_search_fx0 = self.fx_best
+        elif (
+            self.local_solutions
+            and self.evaluator.n_eval - self.last_local_search_neval
+            >= self.local_n2
+        ):
+            quality_order = np.argsort(fx_best_children)
+            # compute minimal distance between the best children and all local
+            #  optima found so far
+            min_distances = np.array(
+                np.min(
+                    np.linalg.norm(y_i - local_solution)
+                    for local_solution in self.local_solutions
+                )
+                for y_i in x_best_children
+            )
+            # sort by furthest distance to existing local optima
+            diversity_order = np.argsort(min_distances)[::-1]
+            # balance quality and diversity (small score is better)
+            priority = (
+                1 - self.balance
+            ) * quality_order + self.balance * diversity_order
+            chosen_child_idx = np.argmin(priority)
+            local_search_x0 = x_best_children[chosen_child_idx]
+            local_search_fx0 = fx_best_children[chosen_child_idx]
+        else:
+            return
+
+        # actual local search
+        # TODO DW: try alternatives if it fails on initial point?
+        optimizer_result: OptimizerResult = self.local_optimizer.minimize(
+            problem=self.evaluator.problem,
+            x0=local_search_x0,
+            id="0",
+        )
+        # add function evaluations during local search to our function
+        #  evaluation counter (NOTE: depending on the setup, we might neglect
+        #  gradient evaluations).
+        self.evaluator.n_eval += optimizer_result.n_fval
+        self.evaluator.n_eval_round += optimizer_result.n_fval
+
+        logger.debug(
+            f"Local search: {local_search_fx0} -> " f"{optimizer_result.fval}"
+        )
+        self.local_solutions.append(optimizer_result.x)
+
+        self._maybe_update_global_best(
+            optimizer_result.x, optimizer_result.fval
+        )
+        self.last_local_search_neval = self.n_iter
+        self.evaluator.reset_round_counter()
+
+    def _maybe_update_global_best(self, x, fx):
+        if fx < self.fx_best:
+            self.x_best = x[:]
+            self.fx_best = fx
+            self.x_best_has_changed = True
+
+    def _go_beyond(self, x_best_children, fx_best_children):
+        """Apply go-beyond strategy.
+
+        See [Egea2009]_ algorithm 1
```

Comment:
In this algorithms, beyond "See xyz", always a short motivation+summary of why+what is happening would be good.